### PR TITLE
SM-14 Reports

### PR DESCRIPTION
**Pull request by olehokhremchuk**
Please review base code and functionality for reports

### CODE REVIEW

File: src/SprykerMiddleware/Zed/Report/Business/Model/ProcessResult.php [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        $spyProcessEntityTransfer->setProcessName($processResultTransfer->getProcessName());
+        $spyProcessEntityTransfer = $this->reportEntityManager->saveProcess($spyProcessEntityTransfer);
+
+        return $this->reportEntityManager->saveProcessResult($processResultTransfer, $spyProcessEntityTransfer);
+    }
+
+    /**
+     * @param int $idResult
+     *
+     * @return mixed
```
Comment (valerio8787) [inline src/SprykerMiddleware/Zed/Report/Business/Model/ProcessResult.php]: Please use  TransferObject instead mixed for return type

File: src/SprykerMiddleware/Zed/Report/Business/Model/ProcessResult.php [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+     */
+    public function findProcessResultByResultId(int $idResult)
+    {
+        return $this->reportRepository->findProcessResultByResultId($idResult);
+    }
+
+    /**
+     * @param int $idProcess
+     *
+     * @return mixed
```
Comment (valerio8787) [inline src/SprykerMiddleware/Zed/Report/Business/Model/ProcessResult.php]: see above

File: src/SprykerMiddleware/Zed/Report/Business/ReportFacade.php [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    {
+        return $this->getFactory()
+            ->createProcessResult()
+            ->findProcessResultByResultId($idResult);
+    }
+
+    /**
+     * @param int $idProcess
+     *
+     * @return mixed
```
Comment (valerio8787) [inline src/SprykerMiddleware/Zed/Report/Business/ReportFacade.php]: Facade should return TO

File: src/SprykerMiddleware/Zed/Report/Business/ReportFacadeInterface.php [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+     * @param int $idResult
+     *
+     * @return \Generated\Shared\Transfer\ProcessResultTransfer
+     */
+    public function findProcessResultByResultId(int $idResult);
+
+    /**
+     * @param int $idProcess
+     *
+     * @return mixed
```
Comment (valerio8787) [inline src/SprykerMiddleware/Zed/Report/Business/ReportFacadeInterface.php]: see above

File: src/SprykerMiddleware/Zed/Report/Communication/Controller/ProcessController.php [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+     * @return array
+     */
+    public function indexAction(Request $request)
+    {
+        $idProcess = $this->castId($request->query->get(static::URL_PARAM_ID_PROCESS));
+        $table = $this->getFactory()->createProcessResultsTable($idProcess);
+
+        $processResult = $this->getFacade()->findProcessByProcessId($idProcess);
+
+        $process['Process Id'] = $processResult->getIdProcess();
```
Comment (valerio8787) [inline src/SprykerMiddleware/Zed/Report/Communication/Controller/ProcessController.php]: @olehokhremchuk Please pass $processResult TO to template as is and process it there

File: src/SprykerMiddleware/Zed/Report/Communication/Controller/ResultController.php [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+     *
+     * @return array
+     */
+    public function indexAction(Request $request)
+    {
+        $idResult = $this->castId($request->query->get(static::URL_PARAM_ID_RESULT));
+
+        $processResult = $this->getFacade()->findProcessResultByResultId($idResult);
+
+        $attributes = [];
```
Comment (valerio8787) [inline src/SprykerMiddleware/Zed/Report/Communication/Controller/ResultController.php]: @olehokhremchuk  See above comment about passing TO to twig, this logic is not needed in controller

File: src/SprykerMiddleware/Zed/Report/Communication/Table/ProcessResultsTable.php [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+    }
+
+    /**
+     * @param \Orm\Zed\Report\Persistence\SpyProcessResult $item
+     *
+     * @return string
+     */
+    protected function getStatus(SpyProcessResult $item)
+    {
+        if ($item->getFailedItemCount() > 0) {
```
Comment (valerio8787) [inline src/SprykerMiddleware/Zed/Report/Communication/Table/ProcessResultsTable.php]: move html to template

File: src/SprykerMiddleware/Zed/Report/Communication/Table/ProcessesTable.php [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    /**
+     * @param array $item
+     *
+     * @return string
+     */
+    protected function createViewButton(array $item)
+    {
+        $viewDiscountUrl = Url::generate(
```
Comment (valerio8787) [inline src/SprykerMiddleware/Zed/Report/Communication/Table/ProcessesTable.php]: Please rename variable

File: src/SprykerMiddleware/Zed/Report/Persistence/Mapper/ProcessResultMapper.php [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+        }
+        return $stagesResultsArray;
+    }
+
+    /**
+     * @param array $stagesResults
+     *
+     * @return \ArrayObject
+     */
+    public function mapArrayToStagesResult(array $stagesResults)
```
Comment (valerio8787) [inline src/SprykerMiddleware/Zed/Report/Persistence/Mapper/ProcessResultMapper.php]: Pass $processResultTransfer to this method too, and use $processResultTransfer->addStageResult(stagesResultsTransfer) instead temp variable

File: src/SprykerMiddleware/Zed/Report/Persistence/ReportEntityManager.php [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+
+/**
+ * @method \SprykerMiddleware\Zed\Report\Persistence\ReportPersistenceFactory getFactory()
+ */
+class ReportEntityManager extends AbstractEntityManager implements ReportEntityManagerInterface, EntityManagerInterface
+{
+    /**
+     * @param \Generated\Shared\Transfer\SpyProcessEntityTransfer $spyProcessEntityTransfer
+     *
+     * @return mixed
```
Comment (valerio8787) [inline src/SprykerMiddleware/Zed/Report/Persistence/ReportEntityManager.php]: TO as return type

File: src/SprykerMiddleware/Zed/Report/Persistence/ReportEntityManagerInterface.php [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+
+use Generated\Shared\Transfer\ProcessResultTransfer;
+use Generated\Shared\Transfer\SpyProcessEntityTransfer;
+
+interface ReportEntityManagerInterface
+{
+    /**
+     * @param \Generated\Shared\Transfer\SpyProcessEntityTransfer $processEntityTransfer
+     *
+     * @return mixed
```
Comment (valerio8787) [inline src/SprykerMiddleware/Zed/Report/Persistence/ReportEntityManagerInterface.php]: Please change mixed to concrete type everywhere

File: src/SprykerMiddleware/Zed/Report/Persistence/ReportPersistenceFactory.php [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+
+    /**
+     * @return \Orm\Zed\Report\Persistence\SpyProcessResultQuery
+     */
+    public function createProcessResultQuery(): SpyProcessResultQuery
+    {
+        return SpyProcessResultQuery::create();
+    }
+
+    public function createProcessResultMapper()
```
Comment (valerio8787) [inline src/SprykerMiddleware/Zed/Report/Persistence/ReportPersistenceFactory.php]: add doc-block and return type

File: src/SprykerMiddleware/Zed/Report/Persistence/ReportPersistenceFactory.php [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+    {
+        return new ProcessResultMapper($this->getUtilEncodingService());
+    }
+
+    /**
+     * @return \SprykerMiddleware\Zed\Report\Dependency\Service\ReportToUtilEncodingInterface
+     */
+    public function getUtilEncodingService()
```
Comment (valerio8787) [inline src/SprykerMiddleware/Zed/Report/Persistence/ReportPersistenceFactory.php]: return type

File: src/SprykerMiddleware/Zed/Report/Persistence/ReportRepository.php [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+ * @method \SprykerMiddleware\Zed\Report\Persistence\ReportPersistenceFactory getFactory()
+ */
+class ReportRepository extends AbstractRepository implements ReportRepositoryInterface
+{
+    /**
+     * @param int $idResult
+     *
+     * @return \Generated\Shared\Transfer\ProcessResultTransfer|null
+     */
+    public function findProcessResultByResultId(int $idResult)
```
Comment (valerio8787) [inline src/SprykerMiddleware/Zed/Report/Persistence/ReportRepository.php]: use return type declaration everywhere

File: src/SprykerMiddleware/Zed/Report/Persistence/ReportPersistenceFactory.php [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+use SprykerMiddleware\Zed\Report\ReportDependencyProvider;
+
+class ReportPersistenceFactory extends AbstractPersistenceFactory
+{
+    /**
+     * @return \Orm\Zed\Report\Persistence\SpyProcessQuery
+     */
+    public function createProcessQuery(): SpyProcessQuery
+    {
+        return SpyProcessQuery::create();
```
Comment (valerio8787) [inline src/SprykerMiddleware/Zed/Report/Persistence/ReportPersistenceFactory.php]: I think you should use query objects that you provide in dependency provider

File: src/SprykerMiddleware/Zed/Report/ReportDependencyProvider.php [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+
+        return $container;
+    }
+
+    /**
+     * @param \Spryker\Zed\Kernel\Container $container
+     *
+     * @return \Spryker\Zed\Kernel\Container
+     */
+    protected function addServiceUtilEncoding(Container $container)
```
Comment (valerio8787) [inline src/SprykerMiddleware/Zed/Report/ReportDependencyProvider.php]: use return type declaration :Container

**Review by valerio8787 — CHANGES_REQUESTED** [review]

**Comment by valerio8787** [comment]
@olehokhremchuk  Also please add declaration of required TO's from ProcessModule

File: src/SprykerMiddleware/Zed/Report/Communication/Table/ProcessResultsTable.php (hunk 7 above)
Comment (olehokhremchuk) [reply src/SprykerMiddleware/Zed/Report/Communication/Table/ProcessResultsTable.php]: @valerio8787 to which template? it's an instance of AbstractTable entity. It has no template, or I should create special template for "label" component and include it to table cell someway?

**Review by valerio8787 — APPROVED** [review]

**Pull request merged**